test/tests/functional/pbs_hook_set_jobenv.py

```diff
@@ -0,0 +1,887 @@
+                         os.environ['TEST_DQUOTE5'] +
+                         " == " + self.env_hook['TEST_DQUOTE5'].rstrip('\n'))
+        self.assertEqual(os.environ['TEST_DQUOTE6'],
+                         self.env_hook['TEST_DQUOTE6'].rstrip('\n'))
+        self.logger.info("TEST_DQUOTE6 matched - " +
+                         os.environ['TEST_DQUOTE6'] +
+                         " == " + self.env_hook['TEST_DQUOTE6'].rstrip('\n'))
+        self.assertEqual(os.environ['TEST_SQUOTE'],
+                         self.env_hook['TEST_SQUOTE'].rstrip('\n'))
+        self.logger.info("TEST_SQUOTE matched - " + os.environ['TEST_SQUOTE'] +
+                         " == " + self.env_hook['TEST_SQUOTE'].rstrip('\n'))
+        self.assertEqual(os.environ['TEST_SQUOTE2'],
+                         self.env_hook['TEST_SQUOTE2'].rstrip('\n'))
+        self.logger.info("TEST_SQUOTE2 matched - " +
+                         os.environ['TEST_SQUOTE2'] +
+                         " == " + self.env_hook['TEST_SQUOTE2'].rstrip('\n'))
+        self.assertEqual(os.environ['TEST_SQUOTE3'],
+                         self.env_hook['TEST_SQUOTE3'].rstrip('\n'))
+        self.logger.info("TEST_SQUOTE3 matched - " +
+                         os.environ['TEST_SQUOTE3'] +
+                         " == " + self.env_hook['TEST_SQUOTE3'].rstrip('\n'))
+        self.assertEqual(os.environ['TEST_SQUOTE4'],
+                         self.env_hook['TEST_SQUOTE4'].rstrip('\n'))
+        self.logger.info("TEST_SQUOTE4 matched - " +
+                         os.environ['TEST_SQUOTE4'] +
+                         " == " + self.env_hook['TEST_SQUOTE4'].rstrip('\n'))
+        self.assertEqual(os.environ['TEST_SQUOTE5'],
+                         self.env_hook['TEST_SQUOTE5'].rstrip('\n'))
+        self.logger.info("TEST_SQUOTE5 matched - " +
+                         os.environ['TEST_SQUOTE5'] +
+                         " == " + self.env_hook['TEST_SQUOTE5'].rstrip('\n'))
+        self.assertEqual(os.environ['TEST_SQUOTE6'],
+                         self.env_hook['TEST_SQUOTE6'].rstrip('\n'))
+        self.logger.info("TEST_SQUOTE6 matched - " +
+                         os.environ['TEST_SQUOTE6'] +
+                         " == " + self.env_hook['TEST_SQUOTE6'].rstrip('\n'))
+        self.assertEqual(os.environ['TEST_SPECIAL'],
+                         self.env_hook['TEST_SPECIAL'].rstrip('\n'))
+        self.logger.info("TEST_SPECIAL matched - " +
+                         os.environ['TEST_SPECIAL'] +
+                         " == " + self.env_hook['TEST_SPECIAL'].rstrip('\n'))
+        self.assertEqual(os.environ['TEST_SPECIAL2'],
+                         self.env_hook['TEST_SPECIAL2'].rstrip('\n'))
+        self.logger.info("TEST_SPECIAL2 matched - " +
+                         os.environ['TEST_SPECIAL2'] +
+                         " == " + self.env_hook['TEST_SPECIAL2'].rstrip('\n'))
+
+    def create_and_submit_job(self, user=None, attribs={}, content=None,
```

Comment:
I really want it defaulted to None so it will execute as the user running the testsuite.

Reply:
What is your reasoning behind wanting these jobs to run as the user running the test suite?  If you run it as TEST_USER, you know how it will run every time you run the test.  If it runs as the person running the test, you might have variances in the user.

Reply:
The testsuite sets the current user's environment variables using the os.environ structure, filling it with special characters that is needed to validate the bug, and then allow qsub -V submission from the current user. If we do it as TEST_USER, PTL would do something like sudo TEST_USER and I'd need need to set/pass the environment variables another way, which could get very tricky.   The test validates a particular set of environment variable values and will not change based on who is executing the testsuite.

Reply:
@bhroam I have not gone though the code but just replying from above conversion, What @bayucan doing is right as per his requirement since as of now PTL does 'sudo -u username qsub' to submit job, and sudo will strip non-allowed (in sudoers file) environment while running qsub.
But I think is nice-to-have feature in PTL and I would request @bayucan to file an issue to allow passing environment while submission in PTL.

Reply:
Thanks @hirenvadalia  To get this PTL testsuite working, I still needed the interactive test with qsub -V -I to work with current user, so I filed PP-1007:
Have an option to pass current env to interactive jobs, and I"m fixing that as part of this pull request. I put in some option that will cause interactive job submission to do: sudo -E <current_user> qsub -I -V
But yes, I'll file another ticket to generalize this to regular, non-interactive jobs and for any test users.

Reply:
@bayucan Thanks for filing ticket. But I doubt that sudo -E will work, because AFAIK even if you use -E still sudo will strip environment if 'Defaults env_reset' is set in sudoers file. But you should try and let us know if it works or not, as it will good information for PP-1007.

Reply:
@hirenvadalia  Interesting....the sudo -E fix works for me probably because I'm running it on my local vmware host, and my sudoers file is not environment restricted. The 2 test cases involving interactive jobs in my testsuite will probably l fail if run in our testbed systems.  So something more general needs to happen to fix PP-1007, or with the fix that I put in, the sudoers file would need to be adjusted on the testbed system.

Reply:
@bayucan Thanks for info. Yup we need to do more general way to fix this issue without depending on sudoers file changes... Anyways... will discuss this more while we work on PP-1007.

Reply:
@hirenvadalia: I have dilemma. As part of this pull request #393, I have the non-general fix to PP-1007. I've added the minor fix to pbs_testlib.py, introducing the 'preserve_end' attribute to job object to allow the '-E' to get passed to qsub -I. Doing this allowed me to run this PTL testsuite in my local test setup that does not restrict sudo environment. Should I just not put in this minor fix to pbs_testlib.py, and keep PP-1007 open, which would require me to just comment out my 2 tests: test_interactive and test_interactive_no_hook? Or should I go ahead and put in my minor fix, make a note of it in PP-1007, leaving the bug open? Or I can just put in my fix in, close PP-1007, and open a new bug, referring to needing a more general fix that the PTL team can work on...